### ISSUE TITLE
Testing and doco of flux-limited TVD discretisation in PorousFlow

### ISSUE THREAD

**Issue by hsheldon** [idaholab/moose#12916]
## Rationale

Flux-limited TVD discretisation has been implemented in PorousFlow and now requires testing in some more realistic scenarios.
See #10426 

## Description

Add tests and documentation of flux-limited TVD scheme

## Impact

Demonstrate that flux-limited TVD provides superior results when compared with conventional unwinding in realistic scenarios. Provide examples showing how to use this approach.

**Comment by WilkAndy** [comment]
Thanks for doing this, @hsheldon .  At least, i hope you're volunteering.  For the next month or so I think we may mucking around with the internals of Flux-limited TVD: (1) to make it faster; (2) to make it work with MPI.  But i don't think the user interface will change at all.

**Comment by hsheldon** [comment]
Yep, I'm volunteering! Progress may be slow and sporadic, but I will do something.

**Comment by WilkAndy** [comment]
Cool!

**Comment by hsheldon** [comment]
Andy, are we still limited to constant velocity scenarios?

**Comment by WilkAndy** [comment]
No, the velocity can also be the Darcy velocity (for single-phase, multi-phase, etc).  For that you use the `PorousFlowAdvectiveFluxCalculator*` UserObjects and the `PorousFlowFluxLimitedTVDAdvection` Kernel.  See

https://mooseframework.inl.gov/modules/porous_flow/kt.html

However, if you want a non-constant velocity without the PorousFlow Darcy stuff, you will have to code it yourself.  I hope that will be quite simple: you just need to follow the example in `AdvectiveFluxCalculatorConstantVelocity`.  You will just need to override the `computeVelocity` method with something like an AuxVariable or a Function.

**Comment by hsheldon** [comment]
Its looking good! This is a 2D thermal convection example. Temperature fixed at top and base, fluid can exit through top boundary but other boundaries impermeable. Note the instability in temperature near top boundary without stabilisation; no instability with KT; and no instability but too much diffusion with full upwinding. This was done with the PorousFlowFullySaturated and PorousFlowUnsaturated action. Nice! Now I'm doing some more exploration with the different flux limiters.
![temp_yflux_time4e11_dt5e10](https://user-images.githubusercontent.com/10649586/52832892-a8195580-312e-11e9-82b3-dc4d98fe97b6.png)

**Comment by WilkAndy** [comment]
Oh, this is so good.  It looks like all that effort was actually worth it!

a

**Comment by hsheldon** [comment]
It is slow, but I'm pretty happy with it!

**Comment by WilkAndy** [comment]
Just use a smaller mesh.  :stuck_out_tongue:    Just kidding, i'm working on that slowness.

**Comment by WilkAndy** [comment]
I can speed up many simulations by a factor of 2 now.  But the code is extremely complicated.  Even though i've tested it quite a bit, i will be super surprised if there aren't bugs.

**Comment by hsheldon** [comment]
Ok, I'll see if I can break it!
I'm wondering how you'd like to see the additional examples/documentation arranged. For now I've put the 2D convection case in porous_flow/examples/2D_thermal_convection. I have added a documentation page in porous_flow/doc/content/modules/porous_flow - this shows how the 2D convection example behaves with different stabilisation techniques. Does that sound ok, or should I be putting the 2D convection example somewhere in porous_flow/test/tests?

**Comment by WilkAndy** [comment]
That organisation sounds perfect, thanks.
